pubsub/stanpubsub/stan.go

```diff
@@ -0,0 +1,538 @@
+	pubsub.DefaultURLMux().RegisterSubscription(Scheme, dd)
+}
+
+// This defaultDialer variable is used only for the test purpose.
+var dd *defaultDialer
+
+// errNotInitialized is an error returned when the topic or subscription is not initialized.
+var errNotInitialized = errors.New("stanpubsub: not initialized")
+
+// defaultDialer dials a default NATS Streaming server based on three required environment variables:
+//	- STAN_SERVER_URL - defines NATS Server Url
+//	- STAN_CLUSTER_ID - defines NATS Streaming Cluster ID
+//	- STAN_CLIENT_ID  - defined NATS Streaming Client ID
+type defaultDialer struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
+
+// OpenTopicURL implements pubsub.TopicURLOpener interface.
+func (o *defaultDialer) OpenTopicURL(ctx context.Context, u *url.URL) (*pubsub.Topic, error) {
+	opener, err := o.defaultConn()
+	if err != nil {
+		return nil, fmt.Errorf("open topic %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenTopicURL(ctx, u)
+}
+
+// OpenSubscriptionURL implements pubsub.SubscriptionURLOpener interface.
+func (o *defaultDialer) OpenSubscriptionURL(ctx context.Context, u *url.URL) (*pubsub.Subscription, error) {
+	opener, err := o.defaultConn()
+	if err != nil {
+		return nil, fmt.Errorf("open subscription %v: failed to open default connection: %v", u, err)
+	}
+	return opener.OpenSubscriptionURL(ctx, u)
+}
+
+func (o *defaultDialer) defaultConn() (*URLOpener, error) {
+	o.init.Do(o.createdOpener)
+	return o.opener, o.err
+}
+
+func (o *defaultDialer) createdOpener() {
+	serverURL := os.Getenv("STAN_SERVER_URL")
+	if serverURL == "" {
+		o.err = errors.New("STAN_SERVER_URL environment variable not set")
+		return
+	}
+	clusterID := os.Getenv("STAN_CLUSTER_ID")
+	if clusterID == "" {
+		o.err = errors.New("STAN_CLUSTER_ID environment variable not set")
+	}
+	clientID := os.Getenv("STAN_CLIENT_ID")
+	if clientID == "" {
+		o.err = errors.New("STAN_CLIENT_ID environment variable not set")
+	}
+	o.connect(serverURL, clusterID, clientID)
+}
+
+func (o *defaultDialer) connect(serverURL, clusterID, clientID string) {
+	conn, err := stan.Connect(clusterID, clientID, stan.NatsURL(serverURL))
+	if err != nil {
+		o.err = fmt.Errorf("failed to dial STAN_SERVER_URL %q: STAN_CLUSTER_ID: %q STAN_CLIENT_ID: %q,  %v", serverURL, clusterID, clientID, err)
+		return
+	}
+	o.opener = &URLOpener{Connection: conn}
+}
+
+// Scheme is the URL scheme stanpubsub registers its URLOpeners under on pubsub.DefaultMux.
+const Scheme = "stan"
+
+// URLOpener opens NATS URLs like "stan://mysubject".
+//
+// The URL host+path is used as the subject.
+//
+// No topic query parameters are supported.
+//
+// Following query parameters are supported for subscription:
+//	- queue 		- sets the name of the queue group.
+//	- durableName	- sets the durable name for the subscription.
+//	- maxInflight 	- sets the maximum number of messages the cluster will send without an ACK.
```

Comment:
I don't get this one, see my comment above about "manualAckMode". Is this only needed when that's turned on?

Reply:
This is an option that  by default was used by NATS Streaming to control the size of messages. 
I've decided that it is a great option to be used in the channel size of `stanpubsub.Subscription`.
It is optional, with a default value equal to `stan.DefaultMaxInflight`.